_posts/2015-04-24-FFI-and-Rust.md

```diff
@@ -0,0 +1,264 @@
+as it talks to itself. To solve this problem, **Rust lets you communicate with C
+APIs at no extra cost while providing strong safety guarantees**.
+
+This is also referred to as Rust's foreign function interface (FFI) and is the
+method by which Rust communicates with other programming languages. In following
+Rust's design principles, this is a **zero cost abstraction** which leverages
+language features such as ownership and borrowing to provide a **safe
+interface**.
+
+In this post we'll explore how Rust's zero-cost FFI and ability to write safe
+abstractions by looking at some examples of interacting with C. C is, however,
+just an example, as we'll also see how Rust can easily talk to languages like
+Python and Ruby just as seamlessly as C.
+
+### Talking to C
+
+First, let's start with an example of calling C code from Rust and then take a
+look at how there is no overhead. Starting off simple, here's a C program which
+will simply double all the input it's given:
+
+```c
+int double_input(int input) {
+    return input * 2;
+}
+```
+
+To call this from Rust, one would write this program:
+
+```rust
+extern crate libc;
+
+extern {
+    fn double_input(input: libc::c_int) -> libc::c_int;
+}
+
+fn main() {
+    let input = 4;
+    let output = unsafe { double_input(input) };
+    println!("{} * 2 = {}", input, output);
+}
+```
+
+And that's it! You can try this out for yourself by [checking out the code on
+GitHub][rust2c] and running `cargo run` from that directory. As you can see
+there is 0 setup necessary to call into a C function and there's no overhead on
+Rust's side when calling it. There are, however, a few subtle aspects of this
+Rust program so let's cover each piece in detail.
+
+[rust2c]: https://github.com/alexcrichton/rust-ffi-examples/tree/master/rust-to-c
+
+First up we see `extern crate libc`. [This crate][libc] provides many useful
+type definitions for FFI bindings when talking with C, and it is necessary
+to ensure that both C and Rust agree on the types crossing the language
+boundary.
+
+[libc]: https://crates.io/crates/libc
+
+This leads us nicely into the next part of the program:
+
+```rust
+extern {
+    fn double_input(input: libc::c_int) -> libc::c_int;
+}
+```
+
+In Rust this is a **declaration** of an externally available function. You can
+think of this along the lines of a C header file. Here's where the compiler
+learns about the inputs and outputs of the function, and you can see above that
+this matches our definition in C. Next up we have the main body of the program:
+
+```rust
+fn main() {
+    let input = 4;
+    let output = unsafe { double_input(input) };
+    println!("{} * 2 = {}", input, output);
+}
+```
+
+We see one of the crucial aspects of FFI in Rust here, the `unsafe` block. The
+compiler knows nothing about the implementation of `double_input`, so it must
+assume that memory unsafety *could* happen in this scenario. This may seem
```

Comment:
Instead of 'memory unsafety could happen' you might embellish a colorful scenario like 'so it must assume that `double_input` may vomit nasal-demons all over your heap, which would clearly be memory-unsafe'.